clustering/common/src/main/java/org/jboss/as/clustering/controller/SubsystemExtension.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2022, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.as.clustering.controller;
+
+import java.util.EnumSet;
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import javax.xml.stream.XMLStreamException;
+
+import org.jboss.as.controller.Extension;
+import org.jboss.as.controller.ExtensionContext;
+import org.jboss.as.controller.PathAddress;
+import org.jboss.as.controller.PersistentResourceXMLDescription;
+import org.jboss.as.controller.parsing.ExtensionParsingContext;
+import org.jboss.as.controller.persistence.SubsystemMarshallingContext;
+import org.jboss.dmr.ModelNode;
+import org.jboss.staxmapper.XMLElementReader;
+import org.jboss.staxmapper.XMLElementWriter;
+import org.jboss.staxmapper.XMLExtendedStreamReader;
+import org.jboss.staxmapper.XMLExtendedStreamWriter;
+
+/**
+ * Generic extension implementation that registers a single subsystem.
+ * @author Paul Ferraro
+ */
+public class SubsystemExtension<S extends Enum<S> & Schema<S>> implements Extension {
+
+    private final String name;
+    private final Model currentModel;
+    private final Supplier<ManagementRegistrar<SubsystemRegistration>> registrarFactory;
+    private final S currentSchema;
+    private final Function<S, XMLElementReader<List<ModelNode>>> readerFactory;
+    private final XMLElementWriter<SubsystemMarshallingContext> writer;
+    private final XMLElementReader<List<ModelNode>> currentReader;
+
+    /**
+     * Constructs a new extension using a reader/writer factory.
+     * @param <P> the parser type
```

Comment:
```suggestion
     * @param <S> the parser type
```

Though, unfortunately Javadoc doesn't allow for that at all here.

```
Generating /tmp/org/jboss/as/clustering/controller/SubsystemExtension.html...
/Users/rhusar/git/wildfly/clustering/common/src/main/java/org/jboss/as/clustering/controller/SubsystemExtension.java:60: error: @param name not found
     * @param <S> the parser type
               ^
```

Reply:
Hmm - that would be a bug in the eclipse javadoc auto-generation.